File: LeetCode/0583. Delete Operation for Two Strings/solution.cpp
/**
 * 1306 / 1306 test cases passed.
 * Runtime: 20 ms
 * Memory Usage: 12.1 MB 
 */
class Solution {
public:
    int minDistance(string word1, string word2) {
        int sz1 = word1.size(), sz2 = word2.size();
        vector<vector<int>> dp(sz1 + 1, vector<int>(sz2 + 1));
    
        int longest = 0;
        for (int i = 1; i <= sz1; i++) {
            for (int j = 1; j <= sz2; j++) {
                if (word1[i - 1] == word2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]); 
                } 
            }
        } 
        return sz1 + sz2 - 2 * dp[sz1][sz2];
    }
};

File: LeetCode/0583. Delete Operation for Two Strings/solution.py
"""
1306 / 1306 test cases passed.
Runtime: 236 ms
Memory Usage: 17.2 MB
"""
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n1, n2 = len(word1), len(word2)
        dp = [[0] * (n2 + 1) for _ in range(n1 + 1)] 
        for i in range(1, n1 + 1):
            for j in range(1, n2 + 1):
                if word1[i - 1] == word2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        return n1 + n2 - 2 * dp[n1][n2]
